From 26f7fce0508a149aefb9e9256f682abe87211760 Mon Sep 17 00:00:00 2001
From: Steve Allewell <steve.allewell@gmail.com>
Date: Sat, 13 Oct 2018 20:11:46 +0100
Subject: Fix for import image failing to find a floss color

Searching for a near color sometimes fails due to the distance from the
available colors to the target color.  This fix provided by Sean Enck
will find the nearest color within a larger target area.
---
 src/FlossScheme.cpp | 16 +++++++++-------
 1 file changed, 9 insertions(+), 7 deletions(-)

diff --git a/src/FlossScheme.cpp b/src/FlossScheme.cpp
index e4cd117..2df85b0 100644
--- a/src/FlossScheme.cpp
+++ b/src/FlossScheme.cpp
@@ -65,21 +65,23 @@ Floss *FlossScheme::find(const QColor &color) const
 {
     QListIterator<Floss *> flossIterator(m_flosses);
 
+    Floss *matched = nullptr;
+    int closest = 100;
+
     while (flossIterator.hasNext()) {
         Floss *floss = flossIterator.next();
         QColor c = floss->color();
 
-        if (c == color) {
-            return floss;
-        }
-
         // the color mapping may not be perfect so search for a near match.
-        if (abs(color.red()-c.red())<2 && abs(color.green()-c.green())<2 && abs(color.blue()-c.blue())<2) {
-            return floss;
+        int distance = abs(color.red()-c.red()) + abs(color.green()-c.green()) + abs(color.blue()-c.blue());
+
+        if (distance < closest) {
+            matched = floss;
+            closest = distance;
         }
     }
 
-    return nullptr;
+    return matched;
 }
 
 
-- 
cgit v1.1

